File: packages/shared-backend/amplify/bedrock-agentcore/runtime/memory_manager.py
import logging
import os
from datetime import datetime, timezone

import boto3

logger = logging.getLogger(__name__)

REGION = os.environ.get("AWS_REGION", "ap-northeast-1")


class MemoryManager:
    def __init__(self, memory_id: str) -> None:
        self.memory_id = memory_id
        self.client = boto3.client("bedrock-agentcore", region_name=REGION)

    def save_conversation(self, actor_id: str, session_id: str, user_message: str, agent_response: str) -> None:
        try:
            self.client.create_event(
                memoryId=self.memory_id,
                actorId=actor_id,
                sessionId=session_id,
                eventTimestamp=datetime.now(timezone.utc).isoformat(),
                payload=[
                    {"conversational": {"content": {"text": user_message}, "role": "USER"}},
                    {"conversational": {"content": {"text": agent_response}, "role": "ASSISTANT"}},
                ],
            )
            logger.info("Memory event saved for actor=%s session=%s", actor_id, session_id)
        except Exception as e:
            logger.error("Failed to save memory event: %s", e)

    def list_events(self, session_id: str, actor_id: str, max_results: int = 20) -> list:
        try:
            response = self.client.list_events(
                memoryId=self.memory_id,
                sessionId=session_id,
                actorId=actor_id,
                maxResults=max_results,
                includePayloads=True,
            )
            return response.get("events", [])
        except Exception as e:
            logger.error("Failed to list memory events: %s", e)
            return []

    def retrieve_memory_records(self, actor_id: str, session_id: str, query: str, max_results: int = 10) -> list:
        namespaces = [
            f"/app/semantic/actors/{actor_id}",
            f"/app/preference/actors/{actor_id}",
            f"/app/summarization/actors/{actor_id}/sessions/{session_id}",
        ]
        results = []
        per_ns = max(1, max_results // len(namespaces))
        for namespace in namespaces:
            try:
                response = self.client.retrieve_memory_records(
                    memoryId=self.memory_id,
                    namespace=namespace,
                    searchCriteria={"searchQuery": query, "topK": per_ns},
                )
                results.extend(response.get("memoryRecordSummaries", []))
            except Exception as e:
                logger.warning("Failed to retrieve memory records for namespace %s: %s", namespace, e)
        return results

    def build_conversation_context(self, session_id: str, actor_id: str, current_query: str) -> dict:
        recent_events = self.list_events(session_id, actor_id)
        relevant_memories = self.retrieve_memory_records(actor_id, session_id, current_query)
        return {"recentEvents": recent_events, "relevantMemories": relevant_memories}

File: packages/shared-backend/amplify/bedrock-agentcore/runtime/src/memory-manager.ts
import {
  BedrockAgentCoreClient,
  CreateEventCommand,
  ListEventsCommand,
  RetrieveMemoryRecordsCommand,
  type Event,
  type MemoryRecordSummary,
  type PayloadType,
} from "@aws-sdk/client-bedrock-agentcore";
import type { MemoryContext } from "./types.js";

const MAX_EVENTS = 20;
const MAX_MEMORY_RECORDS = 10;

export class MemoryManager {
  private readonly client: BedrockAgentCoreClient;

  constructor(
    private readonly memoryId: string,
    region: string,
  ) {
    this.client = new BedrockAgentCoreClient({ region });
  }

  async saveConversation(actorId: string, sessionId: string, query: string, answer: string): Promise<void> {
    const payload: PayloadType[] = [
      {
        conversational: {
          content: { text: query },
          role: "USER",
        },
      },
      {
        conversational: {
          content: { text: answer },
          role: "ASSISTANT",
        },
      },
    ];

    await this.client.send(
      new CreateEventCommand({
        memoryId: this.memoryId,
        actorId,
        sessionId,
        eventTimestamp: new Date(),
        payload,
      }),
    );
  }

  async buildConversationContext(sessionId: string, actorId: string, query: string): Promise<MemoryContext> {
    const recentEvents = await this.listEvents(sessionId, actorId);
    const relevantMemories = await this.retrieveMemoryRecords(actorId, sessionId, query);
    return { recentEvents, relevantMemories };
  }

  private async listEvents(sessionId: string, actorId: string): Promise<Event[]> {
    const response = await this.client.send(
      new ListEventsCommand({
        memoryId: this.memoryId,
        sessionId,
        actorId,
        maxResults: MAX_EVENTS,
        includePayloads: true,
      }),
    );

    if (response.events === undefined) {
      throw new Error("AgentCore Memory did not return events");
    }

    return response.events;
  }

  private async retrieveMemoryRecords(actorId: string, sessionId: string, query: string): Promise<MemoryRecordSummary[]> {
    const namespaces = [
      `/app/semantic/actors/${actorId}`,
      `/app/preference/actors/${actorId}`,
      `/app/summarization/actors/${actorId}/sessions/${sessionId}`,
    ];
    const perNamespace = Math.max(1, Math.floor(MAX_MEMORY_RECORDS / namespaces.length));
    const records: MemoryRecordSummary[] = [];

    for (const namespace of namespaces) {
      const response = await this.client.send(
        new RetrieveMemoryRecordsCommand({
          memoryId: this.memoryId,
          namespace,
          searchCriteria: {
            searchQuery: query,
            topK: perNamespace,
          },
        }),
      );

      if (response.memoryRecordSummaries === undefined) {
        throw new Error(`AgentCore Memory did not return records for ${namespace}`);
      }

      records.push(...response.memoryRecordSummaries);
    }

    return records;
  }
}
